azurekvp/src/kvp.rs

```diff
@@ -0,0 +1,503 @@
+}
+
+impl EmitKVPLayer {
+    pub fn new(file_path: std::path::PathBuf) -> Self {
+        Self { file_path }
+    }
+}
+
+impl<S> Layer<S> for EmitKVPLayer
+where
+    S: Subscriber + for<'lookup> LookupSpan<'lookup>,
+{
+    /// Handles event occurrences within a span by extracting the `msg` field,
+    /// recording the event, and writing the event as key-value pairs (KVP) to a log file.
+    ///
+    /// This method uses the `StringVisitor` to capture the message of the event and
+    /// links the event to the span by writing both span and event data into the same file.
+    ///
+    /// If an `ERROR` level event is encountered, a failure flag is inserted into the span's extensions.
+    ///
+    /// # Example
+    /// ```rust
+    /// use tracing::{event, Level};
+    /// event!(Level::INFO, msg = "Event message");
+    /// ```
+    fn on_event(&self, event: &tracing::Event<'_>, ctx: TracingContext<'_, S>) {
+        let mut event_message = String::new();
+
+        let mut visitor = StringVisitor {
+            string: &mut event_message,
+        };
+
+        event.record(&mut visitor);
+
+        if let Some(span) = ctx.lookup_current() {
+            let mut extensions = span.extensions_mut();
+
+            if event.metadata().level() == &tracing::Level::ERROR {
+                extensions.insert("failure".to_string());
+            }
+
+            let event_instant = MyInstant::now();
+
+            handle_event(&event_message, &span, &self.file_path, event_instant);
+        }
+    }
+
+    /// Called when a new span is created. Records the start time of the span
+    /// and stores it as an extension within the span's context, to be used
+    /// for generating telemetry data for Hyper-V.
+    fn on_new_span(
+        &self,
+        _attrs: &Attributes<'_>,
+        id: &Id,
+        ctx: TracingContext<'_, S>,
+    ) {
+        let start_instant = MyInstant::now();
+        if let Some(span) = ctx.span(id) {
+            span.extensions_mut().insert(start_instant);
+        }
+    }
+
+    /// Called when a span is closed and checks for any logged errors to report.
+    /// This method handles encoding and writing the span data
+    /// (start and end times) to the specified KVP file for Hyper-V telemetry.
+    ///
+    /// # Arguments
+    /// * `id` - The unique identifier for the span.
+    /// * `ctx` - The current tracing context, which can be used to access the span.
+    fn on_close(&self, id: Id, ctx: TracingContext<S>) {
+        if let Some(span) = ctx.span(&id) {
+            let mut status = "success".to_string();
+            let mut event_level = "INFO".to_string();
+
+            if let Some(recorded_status) = span.extensions().get::<String>() {
+                if recorded_status == "failure" {
+                    status = "failure".to_string();
+                    event_level = "ERROR".to_string();
+                }
+            }
+
+            let end_time = SystemTime::now();
```

Comment:
Since this is just grabbing the current time, it seems reasonable to push this call down into the `handle_span` call (unless it's being used to make testing easier, which I didn't spot)